resolver-dns/src/main/java/io/netty/resolver/dns/BiDnsQueryLifecycleObserver.java

```diff
@@ -0,0 +1,96 @@
+package io.netty.resolver.dns;
+
+import io.netty.channel.ChannelFuture;
+import io.netty.handler.codec.dns.DnsQuestion;
+import io.netty.handler.codec.dns.DnsResponseCode;
+import io.netty.util.internal.UnstableApi;
+
+import java.net.InetSocketAddress;
+import java.util.List;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+
+/**
+ * Combines two {@link DnsQueryLifecycleObserver} into a single {@link DnsQueryLifecycleObserver}.
+ */
+@UnstableApi
+public final class BiDnsQueryLifecycleObserver implements DnsQueryLifecycleObserver {
```

Comment:
yeah we are on the same page but I still not get why you would need these extra allocs. What is wrong with create one List when create the Factory, make it unmodifiable and then pass the `same` instance to the newly created Observer all the time ? I don't get why this is not possible or why you think this would cause more allocations then what we have here... Creating a new factory should be very rare and so should be no concern as you usually not create and destroy `DnsResolver`instances frequently.